### ISSUE TITLE
Run the LM head only on supervised positions during training

### ISSUE THREAD

**Issue by aarjunsrinivasan** [huggingface/nanoVLM#210]
During training, targets are  `-100`  for image tokens and the prompt, but forward still projects every position through the LM head before `ignore_index`  discards the masked positions. 

We can instead gather the supervised hidden states before the LM head:

```python
keep = targets != -100
logits = self.decoder.head(hidden_states[keep])
loss = F.cross_entropy(logits, targets[keep])
```

This produces the same loss with numerically equivalent gradients while avoiding the full `[B, T, vocab_size]`  logits tensor.

In my H100 benchmark, this improved throughput by **1.24×** and reduced peak reserved memory by about **5 GiB**.

I'll open a PR with the implementation, correctness test, and benchmark details.
